flanker/addresslib/address.py

```diff
@@ -73,27 +81,92 @@ def parse(address, addr_spec_only=False, metrics=False):
         None
     """
     mtimes = {'parsing': 0}
-
-    parser = flanker.addresslib.parser._AddressParser(False)
+    lexer = flanker.addresslib.lexer.lexer.clone()
+    if addr_spec_only:
+        parser = flanker.addresslib.parser.addr_spec_parser
+    else:
+        parser = flanker.addresslib.parser.mailbox_or_url_parser
+
+    # normalize inputs to bytestrings
+    if isinstance(address, unicode):
+        address = address.encode('utf-8')
+
+    # sanity checks
+    if not address:
+        return None, mtimes
+    if len(address) > MAX_ADDRESS_LENGTH:
+        log.warning('address exceeds maximum length of %s', MAX_ADDRESS_LENGTH)
+        return None, mtimes
 
     try:
-        # addr-spec only
-        if addr_spec_only:
-            bstart = time.time()
-            retval = parser.address_spec(address)
-            mtimes['parsing'] = time.time() - bstart
-            return retval, mtimes
-
-        # full address
         bstart = time.time()
-        retval = parser.address(address)
+        retval = _lift_parser_result(parser.parse(address, lexer=lexer))
         mtimes['parsing'] = time.time() - bstart
-        return retval, mtimes
+    except ply.lex.LexError as e:
+        log.warning(u'error in lexing: %s', e)
+        return None, mtimes
+    except ply.yacc.YaccError as e:
+        log.warning(u'error in parsing: %s', e)
+        return None, mtimes
+    except SyntaxError as e:
+        log.warning(u'error in parsing: %s', e)
+        return None, mtimes
+
+    return retval, mtimes
+
+
+@metrics_wrapper()
+def parse_discrete_list(address_list, metrics=False):
```

Comment:
`parse_list` handles a variety of different input types including a string, a list of strings, and a list of `EmailAddress`es/`UrlAddress`es. There was a lot going on in that method so I encapsulated the logic for parsing a single string into `parse_discrete_list`. It's very similar to `parse` in implementation, just with a different parser entrypoint and a different return type. I struggled to find a good name for `parse_discrete_list` so if you have a better one I'm all ears.